cmake/dts.cmake

```diff
@@ -88,6 +88,63 @@ if(CONFIG_HAS_DTS)
   # preprocessor, and it seems to be including all kinds of
   # directories with who-knows how many header files.
 
+  # Run the C preprocessor on an empty C source file that has one or
+  # more DTS source files -include'd into it to create the
+  # intermediary file *.dts.pre.tmp
+  execute_process(
+    COMMAND ${CMAKE_C_COMPILER}
+    -x assembler-with-cpp
+    -nostdinc
+    -isystem ${DTS_APP_INCLUDE}
+    -isystem ${ZEPHYR_BASE}/include
+    -isystem ${ZEPHYR_BASE}/dts/${ARCH}
+    -isystem ${ZEPHYR_BASE}/dts
+    ${DTC_INCLUDE_FLAG_FOR_DTS}  # include the DTS source and overlays
+    -I${ZEPHYR_BASE}/dts/common
+    ${NOSYSDEF_CFLAG}
+    -D__DTS__
+    -P
+    -E ${ZEPHYR_BASE}/misc/empty_file.c
+    -o ${BOARD}.dts.pre.tmp
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
+  if(NOT "${ret}" STREQUAL "0")
+    message(FATAL_ERROR "command failed with return code: ${ret}")
+  endif()
+
+  # Run the DTC on *.dts.pre.tmp to create the intermediary file *.dts_compiled
+  execute_process(
+    COMMAND ${DTC}
+    -O dts
+    -o ${BOARD}.dts_compiled
+    -b 0
```

Comment:
There is too much copy-paste of code between the first and the second pass.

Copy-paste makes it difficult to read the code.

For instance, is it intentional, or accidental, that only the second pass uses -E unit_address_vs_reg?

To be able to maintain this system, we need to create a CMake function that defines DT passes. The function will be invoked twice and have a prototype that allows us to change how each pass is done.